reconcile/terraform_repo.py

```diff
@@ -0,0 +1,304 @@
+    """
+    Checks whether a Git ref is valid
+    and whether config.tf exists in the project path"""
+    instance = queries.get_gitlab_instance()
+    with GitLabApi(
+        instance, settings=queries.get_secret_reader_settings(), project_url=repo_url
+    ) as gl:
+        try:
+            gl.get_commit_sha(ref=ref, repo_url=repo_url)
+            retrieved_path = gl.get_file("{}/config.tf".format(path), ref)
+
+            if not retrieved_path:
+                raise ParameterError(
+                    'No config.tf found in path: "{}" on repo: "{}". Ensure that a config.tf file is present in this path.'.format(
+                        path, repo_url
+                    )
+                )
+        except (KeyError, AttributeError):
+            raise ParameterError(
+                'Invalid ref: "{}" on repo: "{}"'.format(ref, repo_url)
+            )
+
+
+def diff_missing(
+    a: Mapping[str, TFRepo],
+    b: Mapping[str, TFRepo],
+    requireDelete: bool = False,
+) -> MutableMapping[str, TFRepo]:
+    """
+    Returns a mapping of repos that are present in mapping a but missing in mapping b
+    When requireDelete = True, each repo from a is checked to ensure that the delete flag is set"""
+    missing: MutableMapping[str, TFRepo] = dict()
+    for a_key, a_repo in a.items():
+        b_repo = b.get(a_key)
+        if b_repo is None:
+            if requireDelete is True and a_repo.delete is not True:
+                raise ParameterError(
+                    'To delete the terraform repo "{}", you must set delete: true in the repo definition'.format(
+                        a_repo.name
+                    )
+                )
+            missing[a_key] = a_repo
+
+    return missing
+
+
+def diff_changed(
+    a: Mapping[str, TFRepo], b: Mapping[str, TFRepo]
+) -> MutableMapping[str, TFRepo]:
+    """
+    Returns a mapping of repos that have changed between a and b in the form of returning their b values
+    In order for Terraform to work as expected, the tenant can only update the ref between MRs
+    Updating account, repo, or project_path can lead to unexpected behavior so we error
+    on that
+    """
+    changed: MutableMapping[str, TFRepo] = dict()
+    for a_key, a_repo in a.items():
+        b_repo = b.get(a_key, a_repo)
+        if (
+            a_repo.account != b_repo.account
+            or a_repo.name != b_repo.name
+            or a_repo.project_path != b_repo.project_path
+            or a_repo.repository != b_repo.repository
+        ):
+            raise ParameterError(
+                'Only the `ref` and `delete` parameters for a terraform repo may be updated in merge requests on repo: "{}"'.format(
+                    a_repo.name
+                )
+            )
+        if (a_repo.ref != b_repo.ref) or (a_repo.delete != b_repo.delete):
+            changed[a_key] = b_repo
+
+    return changed
+
+
+def gql_to_tf_repo(repo: TerraformRepoV1) -> TFRepo:
```

Comment:
Is this just for serialization/deserialization? pydantic models can be exported via `model.dict(by_alias=True)` [doc](https://docs.pydantic.dev/usage/exporting_models/), so we can use it directly instead of creating similar classes manually.

Reply:
I was running into issues trying to serialize from JSON to the Pydantic model around https://github.com/app-sre/qontract-reconcile/pull/3464/files#diff-05e8f500c44ae2dc8a3b8087020cc1d2946506df84ea9a118e110c8d8ad82ba2R74

Reply:
However, I will review the docs and take another crack at just using the `TerraformRepoV1` class everywhere instead of `TFRepo` as it is just duplicated classes.

Reply:
Okay I figured that issue out. It was due to the `TerraformRepoV1` class using aliases for some of its fields

Reply:
Now using `TerraformRepoV1` rather than `TFRepo` custom class